src/storage/data_table.cpp

```diff
@@ -169,6 +170,29 @@ DataTable::DataTable(ClientContext &context, DataTable &parent, idx_t removed_co
 	parent.is_root = false;
 }
 
+// Alter column to add new constraint
+DataTable::DataTable(ClientContext &context, DataTable &parent, unique_ptr<Constraint> constraint)
```

Comment:
Can this constructor be removed entirely? A new table should not need to be created, no?

Reply:
When altering table in multiple transaction environmnent, there're 2 different behaviors. 
1. For add_col/rm_col/change_type,  concurrent update to this table is blocked. A new DataTable instance is created and the parent(previous) DT is invalidated (by parent.is_root = false) . 
2. For add_default, concurrent update is allowed as the "default" values only apply to the *future* updates. The target column could contain null values even if it has *default* expr. No new DataTable is created.

Back to set_not_null(or other constraints), although the data is not changed, I think we still need to block other concurrent updates, or other transactions might insert values which violate the constraint. 

```
statement ok con1
begin trainsaction
statement ok con2
begin trainsaction
statement ok con1
alter table t alter column j set not null  // ok as no null values in persistent/local data 
statement ok con2
insert into t(i, j) values(2,null)	// ok as cannot see not_null contraint yet
statement ok con1
commit;
statement ok con2 // now column j is not_null but contains null value!!!
commit;
```

To avoid such issue, I can only think of:
1. Follow the add_col/rm_col/change_type behavior like current implementation, create new DT based on parent DT, the parent is invalidated so concurrent update is blocked. 
2. Or perform the constraint check only during the *commit* time. I'm not sure if it's feasible, and it looks like a little inconsist with other alter command.

I picked the option1 as it's easy (and consist :-)). 
Sorry about the long conversation and maybe I should use other channel for this discussion. Please help me if there's better way to implement this.

Reply:
That's a good point. I agree that this is likely the cleanest way of achieving that. My last question is then did you already test that this works correctly if there are other transactions out there with uncommitted appends that will violate the constraint, e.g.:

```sql
statement ok
CREATE TABLE t(i INTEGER);

statement ok con1
begin transaction

statement ok con2
begin transaction

# con1 adds a NULL value
statement ok con1
INSERT INTO t VALUES (NULL)

# con2 adds a constraint
statement error con2
alter table t alter column i set not null 

```

Reply:
Thanks for your comment!
For concurrent "alter" (con2) and "update" (con1) transactions, the "alter" (con2) always ***Win***. The "update" (con1) will be failed with `"Transaction conflict: adding entries to a table that has been altered!". `Either at the insert statement (if insert after the alter of con2), or at the commit statement (if insert before the alter of con2).
I have tests for above 2 cases in test/sql/alter/alter_col/test_not_null_multi_tran.test: [insertion failure](https://github.com/duckdb/duckdb/pull/4185/files#diff-df085869a02dfb1008a7f51f4c1003f69618fe3a067b20c1d771e0beda2abe07R8-R37), and [commit failure](https://github.com/duckdb/duckdb/pull/4185/files#diff-df085869a02dfb1008a7f51f4c1003f69618fe3a067b20c1d771e0beda2abe07R38-R70).

Reply:
Thanks for your comment!
For concurrent "alter" (con2) and "update" (con1) transactions, the "alter" (con2) always ***Win***. The "update" (con1) will be failed with `"Transaction conflict: adding entries to a table that has been altered!". `Either at the insert statement (if insert after the alter of con2), or at the commit statement (if insert before the alter of con2). And for your test code, con1 will be failed at commit time. 

I have tests for above 2 cases in test/sql/alter/alter_col/test_not_null_multi_tran.test: [insertion failure](https://github.com/duckdb/duckdb/pull/4185/files#diff-df085869a02dfb1008a7f51f4c1003f69618fe3a067b20c1d771e0beda2abe07R8-R37), and [commit failure](https://github.com/duckdb/duckdb/pull/4185/files#diff-df085869a02dfb1008a7f51f4c1003f69618fe3a067b20c1d771e0beda2abe07R38-R70).

But sadly I find another issue which will violate the contraint:
```
statement ok
CREATE TABLE t(i INTEGER);

# con1 begin the tran first
statement ok con1
begin transaction

# con2 begin the tran after con1
statement ok con2
begin transaction

# con1 adds a NULL value
statement ok con1
INSERT INTO t VALUES (NULL)

# con1 commit the update
statement ok con1
COMMIT

# con2 adds a constraint, this should be error, but con2 cannot see the null value (due to MVCC?) 
# it unexpected ***succeeded***
statement error con2
alter table t alter column i set not null 
```
It looks like a write skew abnormal in snapshot isolation, is it an error? Should I fix this?

Reply:
Sorry, yes that is indeed what I meant. That is an error that should be fixed. This can likely be fixed by using `ScanCreateIndex` instead of the regular scan during the constraint verification.

Reply:
Thank you. The `ScanCreateIndex` does fix the issue. However, another unexpected behavior occurs. 
For below test, I thought the constraint can be added as the null value has been deleted locally, but the alter statement failed as it can only see the **Committed** data:
```
statement ok
DROP TABLE IF EXISTS t

statement ok
CREATE TABLE t(i INTEGER, j INTEGER)

statement ok
INSERT INTO t SELECT 1,1 FROM RANGE(1024)

# Insert null
statement ok
INSERT INTO t VALUES(2,NULL)

statement ok
BEGIN TRANSACTION

# Delete null locally
statement ok
DELETE FROM t WHERE i = 2

# We should be able to set the constraint as null has been deleted
# But with ScanCreateIndex (or ScanCommitted), it cannot see the local deletion.
statement ok
ALTER TABLE t ALTER COLUMN j SET NOT NULL

statement ok
COMMIT
```
It seems the regular Scan can merge the local deletion with committed data, but `ScanCreateIndex (or ScanCommitted)`  doesn't. I tried the 3 different TableScanType but all failed. Is there a Scan to find the `committed_data - local_deletion` ?
Anyway, the good thing is this behavior won't violate the constraint, we still have a consist DataTable :-)

Reply:
That does not exist yet, just use the `ScanCreateIndex` and leave the failing test skipped using `mode skip`/`mode unskip`. The constraint being a bit too eager is not a big problem - but letting through invalid values is.

Reply:
Thanks. Just made the change, please check.